{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f062d4d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pyBigWig\n",
    "import h5py\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "# from pyliftover import LiftOver\n",
    "from tqdm.notebook import tqdm\n",
    "from matplotlib.colors import ListedColormap, Normalize\n",
    "import matplotlib.patches as mpatches\n",
    "# from pybedtools import BedTool\n",
    "from collections import defaultdict\n",
    "import pysam\n",
    "from sklearn.metrics import roc_curve, auc\n",
    "from scipy.stats import spearmanr, pearsonr\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "46f72158",
   "metadata": {},
   "outputs": [],
   "source": [
    "# These are the necessary files to run the rest of the notebook, make sure you have them all in the folder\n",
    "\n",
    "sequences_bed_path = 'sequences.bed'\n",
    "\n",
    "shah_tracks = ['GSM5669183_CardiacMyocytes_rep1_LB1.bigwig',\n",
    "               'GSM5669188_APS_rep1_LB1.bigwig',\n",
    "               'GSM5669193_BorderEctoderm_rep1_LB1.bigwig',\n",
    "               'GSM5669198_D5Midbrain_rep1_LB1.bigwig',\n",
    "               'GSM5669203_DefEctoderm_rep1_LB1.bigwig',\n",
    "               'GSM5669209_EarlySomite_rep1_LB1.bigwig',\n",
    "               'GSM5669214_EndoProgenitor_rep1_LB1.bigwig',\n",
    "               'GSM5669220_Epicardium_rep1_LB1.bigwig',\n",
    "               'GSM5669226_H9ESC_rep1_LB1.bigwig',\n",
    "               'GSM5669232_Liver_rep1_LB1.bigwig',\n",
    "               'GSM5669238_MidHindgut_rep1_LB1.bigwig',\n",
    "               'GSM5669244_ParaxMesoderm_rep1_LB1.bigwig']\n",
    "dimelo_tracks = ['HEK293_LMNB1','GM12878_LMNB1','HEK293_CpG_methylation','HEK293_CpG_motifs','GM12878_CpG_methylation','GM12878_CpG_motifs']\n",
    "\n",
    "all_tracks = [shah_track[0:-16] for shah_track in shah_tracks] + dimelo_tracks\n",
    "\n",
    "# /clusterfs/nilah/oberon/datasets/cs282a/\n",
    "dataset_filepath = '/clusterfs/nilah/oberon/datasets/cs282a/dataset_14-lmnb1_4-cpg.h5'\n",
    "\n",
    "cm_LAD_tabix_str = 'GSM5669183_CardiacMyocytes_LADs.sorted.bed.gz'\n",
    "fl_LAD_tabix_str = '/clusterfs/nilah/oberon/datasets/lamina_association_shah/GSM5669232_Liver_LADs.sorted.bed.gz'\n",
    "\n",
    "# /clusterfs/nilah/oberon/datasets/cs282a/inference/\n",
    "linear_probe_inference = '/clusterfs/nilah/oberon/datasets/cs282a/inference/probe_first_full_run.h5'\n",
    "multilayer_inference = 'mlp_model_20231128_063541_2.h5'\n",
    "maxpool_inference = 'maxpool_model_20231128_072156_3.h5'\n",
    "transformer_inference = 'transformer_model_20231128_080512_7.h5'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "84142e4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "h5_files_dict = {\n",
    "    'linear-probe':(linear_probe_inference,'128bp_bins'),\n",
    "    'conv-perceptron':(multilayer_inference,'single_bin'),\n",
    "    'maxpool-perceptron':(maxpool_inference,'single_bin'),\n",
    "    'transformer':(transformer_inference,'single_bin'),\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ad950af8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# These are approximate chromosome lengths from the hg38 reference genome\n",
    "chrom_lengths = {    \n",
    "    'chr1':249000000,\n",
    "    'chr2':243000000,\n",
    "    'chr3':198000000,\n",
    "    'chr4':191000000,\n",
    "    'chr5':182000000,\n",
    "    'chr6':171000000,\n",
    "    'chr7':160000000,\n",
    "    'chr8':146000000,\n",
    "    'chr9':140000000,\n",
    "    'chr10':135000000,\n",
    "    'chr11':135000000,\n",
    "    'chr12':133000000,\n",
    "    'chr13':113500000,\n",
    "    'chr14':107000000,\n",
    "    'chr15':102000000,\n",
    "    'chr16':90000000,\n",
    "    'chr17':84000000,\n",
    "    'chr18':80000000,\n",
    "    'chr19':59000000,\n",
    "    'chr20':66000000,\n",
    "    'chr21':48000000,\n",
    "    'chr22':51000000,\n",
    "    'chrX':155000000,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd05f5c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ad191524d3c14cf2ba39b904ee033476",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/23 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "99c59c6cc19a4f71b593e0603e736b80",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "reading labels and predictions for chr1:   0%|          | 0/38171 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "061d630657d047478d337faa668e9719",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "reading labels and predictions for chr2:   0%|          | 0/38171 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cm_LADs_tabix = pysam.TabixFile(cm_LAD_tabix_str)\n",
    "fl_LADs_tabix = pysam.TabixFile(fl_LAD_tabix_str)\n",
    "# If you have memory limitations, you may prefer to run a subset of chromosomes. \n",
    "# chr14 (element 13 of the list) is almost entirely test set so that is a good choice for assessing\n",
    "plot_chroms = list(chrom_lengths.keys())\n",
    "chrom_results = {}\n",
    "for plot_chrom in tqdm(plot_chroms):\n",
    "    plot_startco = 0\n",
    "    plot_endco = chrom_lengths[plot_chrom]\n",
    "    plot_xcoord = np.arange(start=plot_startco,stop=plot_endco,step=128)\n",
    "    length = len(plot_xcoord)\n",
    "    h5py_files = {}\n",
    "    default_array = np.full((length, 18), np.nan,dtype=float)\n",
    "    def default_factory():\n",
    "        return default_array.copy()\n",
    "    output_arrays = defaultdict(default_factory)\n",
    "    plot_array_low_res = np.full((length, 18), np.nan)\n",
    "    plot_array_high_res = np.full((length, 18), np.nan)\n",
    "    cm_lads_array = np.zeros(length)\n",
    "    fl_lads_array = np.zeros(length)\n",
    "    test_train_split_array = np.zeros(length)\n",
    "\n",
    "\n",
    "    buffer_at_ends = 8192\n",
    "    seqbedlen = 38171\n",
    "    test_train_split_array\n",
    "    with open(sequences_bed_path,'r') as bed_file, h5py.File(dataset_filepath,'r') as f:\n",
    "        for key, file_path in h5_files_dict.items():\n",
    "            h5py_files[key]=h5py.File(file_path[0],'r')\n",
    "\n",
    "        for index,bed_line in tqdm(enumerate(bed_file),total=seqbedlen,desc=f'reading labels and predictions for {plot_chrom}'):\n",
    "            location_fields = bed_line.split('\\t')\n",
    "            chrom = str(location_fields[0])\n",
    "            start = int(location_fields[1])+buffer_at_ends\n",
    "            end = int(location_fields[2])-buffer_at_ends  \n",
    "            set_int = 0\n",
    "            if location_fields[3].strip()=='train':\n",
    "                set_int = 1\n",
    "            elif location_fields[3].strip()=='valid':\n",
    "                set_int = 2\n",
    "            elif location_fields[3].strip()=='test':\n",
    "                set_int = 3\n",
    "\n",
    "            cm_lad_type = 0\n",
    "            fl_lad_type = 0\n",
    "            for interval in cm_LADs_tabix.fetch(plot_chrom,start,end):\n",
    "                fields = interval.split('\\t')\n",
    "                cm_lad_type = max(int(fields[3][1]),cm_lad_type)\n",
    "                \n",
    "            for interval in fl_LADs_tabix.fetch(plot_chrom,start,end):\n",
    "                fields = interval.split('\\t')\n",
    "                fl_lad_type = max(int(fields[3][1]),fl_lad_type)\n",
    "\n",
    "            if end>plot_startco and start<plot_endco and chrom==plot_chrom:\n",
    "                if start>plot_startco and end<plot_endco:\n",
    "                    for key,file_details in h5_files_dict.items():\n",
    "                        output_arrays[key][(start-plot_startco)//128:(end-plot_startco)//128,:] = h5py_files[key][file_details[1]][index,:,:]\n",
    "                    plot_array_low_res[(start-plot_startco)//128:(end-plot_startco)//128,:] = f['single_bin'][index,:,:]\n",
    "                    plot_array_high_res[(start-plot_startco)//128:(end-plot_startco)//128,:] = f['128bp_bins'][index,:,:]\n",
    "                    cm_lads_array[(start-plot_startco)//128:(end-plot_startco)//128] = cm_lad_type\n",
    "                    fl_lads_array[(start-plot_startco)//128:(end-plot_startco)//128] = fl_lad_type\n",
    "                    test_train_split_array[(start-plot_startco)//128:(end-plot_startco)//128] = set_int\n",
    "    chrom_results[plot_chrom] = (\n",
    "        plot_xcoord,\n",
    "        plot_array_low_res,\n",
    "        plot_array_high_res,\n",
    "        output_arrays,\n",
    "        cm_lads_array,\n",
    "        fl_lads_array,\n",
    "        test_train_split_array,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2441de87",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This cell simply spits out the composition of each chromosome in terms of test/train/validation split, as \n",
    "# well as unindexed portions (mostly repetitive regions that are left out of training, e.g. centromeres)\n",
    "sets_strings_dict = {}\n",
    "for plot_chrom in plot_chroms:\n",
    "    (\n",
    "        plot_xcoord,\n",
    "        plot_array_low_res,\n",
    "        plot_array_high_res,\n",
    "        output_arrays,\n",
    "        cm_lads_array,\n",
    "        fl_lads_array,\n",
    "        test_train_split_array,\n",
    "    ) = chrom_results[plot_chrom]\n",
    "    unindexed_num = np.sum(test_train_split_array==0)\n",
    "    train_num = np.sum(test_train_split_array==1)\n",
    "    valid_num = np.sum(test_train_split_array==2)\n",
    "    test_num = np.sum(test_train_split_array==3)\n",
    "    total_num = train_num + valid_num + test_num + unindexed_num\n",
    "    sets_string = f'{100*train_num/total_num:.2f}% training set, {100*valid_num/total_num:.2f}% validation set, {100*test_num/total_num:.2f}% test set, {100*unindexed_num/total_num:.2f}% unindexed'\n",
    "    print(plot_chrom,sets_string)\n",
    "    sets_strings_dict[plot_chrom] = sets_string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3d35544",
   "metadata": {},
   "outputs": [],
   "source": [
    "set_dict = {\n",
    "    'train':1,\n",
    "    'valid':2,\n",
    "    'test':3,\n",
    "}\n",
    "plot_arrays_low_res_list = []\n",
    "plot_arrays_high_res_list = []\n",
    "output_arrays_lists_dict = defaultdict(list)\n",
    "cm_lads_arrays_list = []\n",
    "fl_lads_arrays_list = []\n",
    "test_train_split_arrays_list = []\n",
    "for plot_chrom in plot_chroms:\n",
    "    (\n",
    "    plot_xcoord,\n",
    "    plot_array_low_res,\n",
    "    plot_array_high_res,\n",
    "    output_arrays,\n",
    "    cm_lads_array,\n",
    "    fl_lads_array,\n",
    "    test_train_split_array,\n",
    "    ) = chrom_results[plot_chrom]\n",
    "    plot_arrays_low_res_list.append(plot_array_low_res)\n",
    "    plot_arrays_high_res_list.append(plot_array_high_res)\n",
    "    for key,value in output_arrays.items():\n",
    "        output_arrays_lists_dict[key].append(value)\n",
    "    cm_lads_arrays_list.append(cm_lads_array)\n",
    "    fl_lads_arrays_list.append(fl_lads_array)\n",
    "    test_train_split_arrays_list.append(test_train_split_array)\n",
    "print('concatenating chromosome arrays')\n",
    "plot_array_low_res_concat = np.concatenate(plot_arrays_low_res_list,axis=0)\n",
    "plot_array_high_res_concat = np.concatenate(plot_arrays_high_res_list,axis=0)\n",
    "output_arrays_concat = {}\n",
    "for key,value in output_arrays_lists_dict.items():\n",
    "    output_arrays_concat[key] = np.concatenate(value,axis=0)\n",
    "cm_lads_array_concat = np.concatenate(cm_lads_arrays_list,axis=0)\n",
    "fl_lads_array_concat = np.concatenate(fl_lads_arrays_list,axis=0)\n",
    "test_train_split_array_concat = np.concatenate(test_train_split_arrays_list,axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "727e94a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "scaling_by_track = [1/128,1/128,1/128,1/128,1/128,1/128,1/128,1/128,1/128,1/128,1/128,1/128,1/500,1/500,1/500,1,1/500,1]\n",
    "data_label_by_track = ['(aligned read count)',\n",
    "                       '(aligned read count)',\n",
    "                       '(aligned read count)',\n",
    "                       '(aligned read count)',\n",
    "                       '(aligned read count)',\n",
    "                       '(aligned read count)',\n",
    "                       '(aligned read count)',\n",
    "                       '(aligned read count)',\n",
    "                       '(aligned read count)',\n",
    "                       '(aligned read count)',\n",
    "                       '(aligned read count)',\n",
    "                       '(aligned read count)',\n",
    "                       '(methylated adenines/adenines)',\n",
    "                       '(methylated adenines/adenines)',\n",
    "                       '(methylated CpGs/CpGs)',\n",
    "                       '(CpGs count)',\n",
    "                       '(methylated CpGs/CpGs)',\n",
    "                       '(CpGs count)'\n",
    "                      ]\n",
    "for i,lads_array in [(0,cm_lads_array_concat),(9,fl_lads_array_concat)]:\n",
    "    for subset_name,subset_label  in set_dict.items():\n",
    "        nas_i = np.isnan(plot_array_low_res_concat[:,i])\n",
    "        subset_mask = (test_train_split_array_concat==subset_label)\n",
    "        regions_count = np.sum(subset_mask & ~nas_i)\n",
    "\n",
    "        fig1, axs1 = plt.subplots(1,5,figsize=(24,5))\n",
    "        fig1.suptitle(f'ROC for predicting lamina association domains\\n{all_tracks[i]} {subset_name}, {plot_chroms[0]} to {plot_chroms[-1]}, {regions_count} 128bp regions',fontweight='bold',size=22)\n",
    "        # Calculate ROC curve from labels\n",
    "        fpr, tpr, thresholds = roc_curve(lads_array[subset_mask]>0,np.nan_to_num(plot_array_high_res_concat[subset_mask,0]))\n",
    "        roc_auc = auc(fpr, tpr)\n",
    "        axs1[0].plot(fpr, tpr, color='darkorange', lw=2, label=f'ROC curve (area = %0.4f)' % roc_auc)\n",
    "        axs1[0].plot([0, 1], [0, 1], color='navy', lw=2, linestyle='--')\n",
    "        axs1[0].set_xlim([0.0, 1.0])\n",
    "        axs1[0].set_ylim([0.0, 1.05])\n",
    "        axs1[0].set_xlabel('False Positive Rate: Incorrectly calling LAD')\n",
    "        axs1[0].set_ylabel('True Positive Rate: Correctly calling LAD')\n",
    "        axs1[0].set_title(f'from labels')    \n",
    "        axs1[0].legend(loc=\"lower right\")\n",
    "\n",
    "        for indx,(key,output_array) in enumerate(output_arrays_concat.items()):\n",
    "            fpr, tpr, thresholds = roc_curve(lads_array[subset_mask]>0,np.nan_to_num(output_array[subset_mask,0]))\n",
    "            roc_auc = auc(fpr, tpr)\n",
    "\n",
    "            axs1[indx+1].plot(fpr, tpr, color='darkorange', lw=2, label='ROC curve (area = %0.4f)' % roc_auc)\n",
    "            axs1[indx+1].plot([0, 1], [0, 1], color='navy', lw=2, linestyle='--')\n",
    "            axs1[indx+1].set_xlim([0.0, 1.0])\n",
    "            axs1[indx+1].set_ylim([0.0, 1.05])\n",
    "            axs1[indx+1].set_xlabel('False Positive Rate: Incorrectly calling LAD')\n",
    "            axs1[indx+1].set_ylabel('True Positive Rate: Correctly calling LAD')\n",
    "            axs1[indx+1].set_title(f'{key}')\n",
    "            axs1[indx+1].legend(loc=\"lower right\")\n",
    "        fig1.subplots_adjust(left=0.1, right=0.9, top=0.8, bottom=0.1, hspace=0.5, wspace=0.4)\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abb7c7aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subset_name,subset_label  in set_dict.items():\n",
    "    nas_i = np.isnan(plot_array_low_res_concat[:,0])\n",
    "    subset_mask = (test_train_split_array_concat==subset_label)\n",
    "    regions_count = np.sum(subset_mask & ~nas_i)\n",
    "    fig1, axs1 = plt.subplots(1,4,figsize=(24,5))\n",
    "    fig1.suptitle(f'ROC for predicting differential normalized LMNB1 signal\\n{all_tracks[0]} vs {all_tracks[9]} {subset_name}, {plot_chroms[0]} to {plot_chroms[-1]}, {regions_count} 128bp regions',fontweight='bold',size=22)\n",
    "    mean_cm = np.nanmean(plot_array_low_res_concat[:,0])\n",
    "    std_cm = np.nanstd(plot_array_low_res_concat[:,0])\n",
    "    mean_fl = np.nanmean(plot_array_low_res_concat[:,9])\n",
    "    std_fl = np.nanstd(plot_array_low_res_concat[:,9])\n",
    "\n",
    "    meas_differences = ((plot_array_low_res_concat[:,9]-mean_fl)/std_fl-(plot_array_low_res_concat[:,0]-mean_cm)/std_cm)\n",
    "    \n",
    "    for indx,(key,output_array) in enumerate(output_arrays_concat.items()):\n",
    "        pred_differences = ((output_array[:,9]-mean_fl)/std_fl-(output_array[:,0]-mean_cm)/std_cm)\n",
    "        fpr1, tpr1, thresholds1 = roc_curve(np.nan_to_num(meas_differences>0)[subset_mask],np.nan_to_num(pred_differences[subset_mask]))\n",
    "        roc_auc1 = auc(fpr1, tpr1)\n",
    "        axs1[indx].plot(fpr1, tpr1, color='darkorange', lw=2, label=f'stronger in myocyte (area = %0.4f)' % roc_auc1)\n",
    "        axs1[indx].plot([0, 1], [0, 1], color='navy', lw=2, linestyle='--')\n",
    "        axs1[indx].set_xlim([0.0, 1.0])\n",
    "        axs1[indx].set_ylim([0.0, 1.05])\n",
    "        axs1[indx].set_xlabel('False Positive Rate: Incorrectly calling Myocyte>Liver')\n",
    "        axs1[indx].set_ylabel('True Positive Rate: Correctly calling Myocyte>Liver')\n",
    "        axs1[indx].set_title(f'{key}')    \n",
    "        axs1[indx].legend(loc=\"lower right\")   \n",
    "    fig1.subplots_adjust(left=0.1, right=0.9, top=0.8, bottom=0.1, hspace=0.5, wspace=0.4)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd273306",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subset_name,subset_label  in set_dict.items():\n",
    "    nas_i = np.isnan(plot_array_low_res_concat[:,0])\n",
    "    subset_mask = (test_train_split_array_concat==subset_label)\n",
    "    regions_count = np.sum(subset_mask & ~nas_i)\n",
    "    fig1, axs1 = plt.subplots(1,4,figsize=(24,5))\n",
    "    fig1.suptitle(f'ROC for predicting differential CpG methylation\\n{all_tracks[14]} vs {all_tracks[16]} {subset_name}, {plot_chroms[0]} to {plot_chroms[-1]}, {regions_count} 128bp regions',fontweight='bold',size=22)\n",
    "\n",
    "    meas_differences = plot_array_low_res_concat[:,14]-plot_array_low_res_concat[:,16]\n",
    "    \n",
    "    for indx,(key,output_array) in enumerate(output_arrays_concat.items()):\n",
    "        pred_differences = output_array[:,14]-output_array[:,16]\n",
    "        fpr1, tpr1, thresholds1 = roc_curve(np.nan_to_num(meas_differences>0)[subset_mask],np.nan_to_num(pred_differences[subset_mask]))\n",
    "        roc_auc1 = auc(fpr1, tpr1)\n",
    "        axs1[indx].plot(fpr1, tpr1, color='darkorange', lw=2, label=f'stronger in HEK (area = %0.4f)' % roc_auc1)\n",
    "        axs1[indx].plot([0, 1], [0, 1], color='navy', lw=2, linestyle='--')\n",
    "        axs1[indx].set_xlim([0.0, 1.0])\n",
    "        axs1[indx].set_ylim([0.0, 1.05])\n",
    "        axs1[indx].set_xlabel('False Positive Rate: Incorrectly calling HEK293>GM12878')\n",
    "        axs1[indx].set_ylabel('True Positive Rate: Correctly calling HEK293>GM12878')\n",
    "        axs1[indx].set_title(f'{key}')    \n",
    "        axs1[indx].legend(loc=\"lower right\")   \n",
    "    fig1.subplots_adjust(left=0.1, right=0.9, top=0.8, bottom=0.1, hspace=0.5, wspace=0.4)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8d3f727",
   "metadata": {},
   "outputs": [],
   "source": [
    "colors = {0: '1.0', 1: '0.9', 2: '0.75', 3: '0.5'}\n",
    "def apply_fill_between(ax, x_coords, y_data, split_array, colors):\n",
    "    start_idx = 0\n",
    "    for i in range(1, len(split_array)):\n",
    "        if split_array[i] != split_array[start_idx]:\n",
    "            ax.fill_between(x=x_coords[start_idx:i], y1=y_data[start_idx:i], color=colors[split_array[start_idx]])\n",
    "            start_idx = i\n",
    "    \n",
    "    ax.fill_between(x_coords[start_idx:], y_data[start_idx:], color=colors[split_array[start_idx]])\n",
    "\n",
    "\n",
    "for plot_chrom in plot_chroms:\n",
    "    (\n",
    "    plot_xcoord,\n",
    "    plot_array_low_res,\n",
    "    plot_array_high_res,\n",
    "    output_arrays,\n",
    "    cm_lads_array,\n",
    "    fl_lads_array,\n",
    "    test_train_split_array,\n",
    "    ) = chrom_results[plot_chrom]\n",
    "    for i,track_name in enumerate(all_tracks):\n",
    "        scaling = scaling_by_track[i]\n",
    "        axlabel = data_label_by_track[i]\n",
    "        fig1, axs1 = plt.subplots(2,4,figsize=(20,6))\n",
    "        fig1.suptitle(f'Chromosome profiles for {track_name} data, {plot_chrom}: {sets_strings_dict[plot_chrom]}',fontweight='bold',size=18)\n",
    "        for idx,(key,output_array) in enumerate(output_arrays.items()):\n",
    "            if h5_files_dict[key][1]=='128bp_bins':\n",
    "                axs1[0,idx].plot(plot_xcoord,plot_array_high_res[:,i]*scaling,'k',label='128bp bin labels')  \n",
    "                axs1[1,idx].plot(plot_xcoord,output_array[:,i]*scaling,'r-',alpha=0.6,label='128bp bin predictions')\n",
    "            elif h5_files_dict[key][1]=='single_bin':\n",
    "                axs1[0,idx].plot(plot_xcoord,plot_array_low_res[:,i]*scaling,'k',label='114kbp bin labels')\n",
    "                axs1[1,idx].plot(plot_xcoord,output_array[:,i]*scaling,'r-',alpha=0.6,label='114kbp bin predictions')\n",
    "            axs1[0,idx].set_title(f'labels {key}')\n",
    "            axs1[1,idx].set_title(f'predictions {key}')\n",
    "            axs1[0,idx].set_ylabel(f'labels {axlabel}')\n",
    "            axs1[1,idx].set_ylabel(f'predictions {axlabel}')\n",
    "            axs1[0,idx].set_xlabel(f'{plot_chrom} coordinate')\n",
    "            axs1[1,idx].set_xlabel(f'{plot_chrom} coordinate')\n",
    "#             axs1[idx].legend(loc=\"upper center\")\n",
    "        plt.tight_layout()\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2154bbca",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "basenji",
   "language": "python",
   "name": "basenji"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
