{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-07-09 11:39:34.102280: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /home/drk/code/torch_1-2018/install/lib:/usr/local/cuda-11.2/targets/x86_64-linux/lib:\n",
      "2022-07-09 11:39:34.102336: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n"
     ]
    }
   ],
   "source": [
    "import glob\n",
    "import gzip\n",
    "import json\n",
    "from natsort import natsorted\n",
    "import os\n",
    "import shutil\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pysam\n",
    "import seaborn as sns\n",
    "import tensorflow as tf\n",
    "\n",
    "import pygene\n",
    "from basenji import dna_io\n",
    "from basenji import metrics\n",
    "from basenji import trainer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(44)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Precursors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# human\n",
    "stability_file = '~vagar/RNAbinding/forPaper/Fig4/binnedgenes.txt'\n",
    "vikram_gtf_file = '/home/vagar/RNAbinding/forPaper/datasets/human/Homo_sapiens.GRCh38.83.chosenTranscript.gtf'\n",
    "# human_gtf_file = '/home/vagar/RNAbinding/forPaper/datasets/human/Homo_sapiens.GRCh38.83.chosenTranscript.gtf'\n",
    "genome_fasta_file = '%s/assembly/ensembl/Homo_sapiens.GRCh38.dna.primary_assembly.fa' % os.environ['HG38']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # mouse\n",
    "# stability_file = 'mouse_stability.tsv'\n",
    "# vikram_gtf_file = '/home/vagar/RNAbinding/forPaper/datasets/mouse/Mus_musculus.GRCm3.90.chosenTranscript.gtf.gz'\n",
    "# genome_fasta_file = '%s/assembly/ensembl/Mus_musculus.GRCm38.dna.primary_assembly.fa' % os.environ['MM10']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def trim_dot(gene_id):\n",
    "    dot_i = gene_id.find('.')\n",
    "    if dot_i != -1:\n",
    "        gene_id = gene_id[:dot_i]\n",
    "    return gene_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_gene_fasta(fasta_file, trim_geneid=False):\n",
    "    if fasta_file[-3:] == '.gz':\n",
    "        fasta_open = gzip.open(fasta_file, 'rt')\n",
    "    else:\n",
    "        fasta_open = open(fasta_file, 'r')\n",
    "        \n",
    "    gene_seq = {}\n",
    "    for line in fasta_open:\n",
    "        if line[0] == '>':\n",
    "            gene = line[1:].rstrip()\n",
    "            if trim_geneid:\n",
    "                gene = trim_dot(gene)\n",
    "            gene_seq[gene] = ''\n",
    "        else:\n",
    "            gene_seq[gene] += line.rstrip()\n",
    "            \n",
    "    fasta_open.close()\n",
    "        \n",
    "    return gene_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rna_1hot_splice(transcript, genome_open):\n",
    "    \"\"\"Extract RNA nucleotides and add splice and coding frame channels.\n",
    "        Consider strand, but do not yet reverse complement, because its\n",
    "        easier to handle variant substitution.\"\"\"\n",
    "\n",
    "    # exons RNA\n",
    "    tx_exons = transcript.fasta_exons(genome_open, stranded=False)\n",
    "    seq_len = len(tx_exons)\n",
    "\n",
    "    # CDS RNA\n",
    "    tx_cds = transcript.fasta_cds(genome_open, stranded=False)\n",
    "    if len(tx_cds) % 3 == 0:\n",
    "        valid_coding = True\n",
    "    else:\n",
    "        # print('WARNING: %d mod 3 == %d length:' % (len(tx_cds),len(tx_cds)%3), transcript)\n",
    "        valid_coding = False\n",
    "    cds_start = tx_exons.find(tx_cds)\n",
    "    if cds_start == -1:\n",
    "        print(transcript)\n",
    "    cds_end = cds_start + len(tx_cds)\n",
    "\n",
    "    # add frame channel\n",
    "    frame_channel = np.zeros((seq_len,1))\n",
    "    if valid_coding:\n",
    "        aa_len = (cds_end - cds_start) // 3\n",
    "        if transcript.strand == '+':\n",
    "            frame_channel[cds_start:cds_end,0] = np.tile([1,0,0], aa_len)\n",
    "        else:\n",
    "            frame_channel[cds_start:cds_end,0] = np.tile([0,0,1], aa_len)\n",
    "\n",
    "    # splice track\n",
    "    splice_channel = np.zeros((seq_len,1))\n",
    "    ti = 0\n",
    "    exon_lens = [ex.end-ex.start for ex in transcript.exons]\n",
    "    for el in exon_lens:\n",
    "        if transcript.strand == '+':\n",
    "            ti += el\n",
    "            splice_channel[ti-1,0] = 1\n",
    "        else:\n",
    "            splice_channel[ti,0] = 1\n",
    "            ti += el\n",
    "\n",
    "    # 1 hot encode\n",
    "    tx_seq1 = dna_io.dna_1hot(tx_exons, n_sample=True)\n",
    "    tx_seq1 = np.concatenate([tx_seq1, frame_channel, splice_channel], axis=-1)\n",
    "    \n",
    "    # reverse complement\n",
    "    if transcript.strand == '-':\n",
    "        tx_seq1 = dna_io.hot1_rc(tx_seq1)\n",
    "\n",
    "    return tx_seq1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_rna_length(tx_seq1, length):\n",
    "    \"\"\"Set RNA length to specific value by padding or trimming.\"\"\"\n",
    "    seq_len, seq_depth = tx_seq1.shape\n",
    "\n",
    "    if seq_len < length:\n",
    "        # pad\n",
    "        pad_len = length - seq_len\n",
    "        pad_seq1 = np.zeros((pad_len,seq_depth))\n",
    "        tx_seq1 = np.concatenate([tx_seq1,pad_seq1], axis=0)\n",
    "\n",
    "    elif seq_len > length:\n",
    "        # trim 5'\n",
    "        tx_seq1 = tx_seq1[-length:]\n",
    "\n",
    "    return tx_seq1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read stability measurements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read sample table\n",
    "stability_df = pd.read_csv(stability_file, sep='\\t')\n",
    "stability_df.columns = ['Fold','Gene','Stability']\n",
    "stability_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_gtf_file = 'genes.gtf'\n",
    "genes_gtf_out = open(genes_gtf_file, 'w')\n",
    "    \n",
    "for line in open(vikram_gtf_file):\n",
    "    a = line.rstrip().split('\\t')\n",
    "    tx_id, gene_id = a[-1].split(':')\n",
    "    a[-1] = 'gene_id \"%s\"; transcript_id \"%s\"; saluki_score \"1.0\";' % (gene_id, tx_id)\n",
    "    print('\\t'.join(a), file=genes_gtf_out)\n",
    "        \n",
    "genes_gtf_out.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 + gene_id \"ENSG00000122484\"; transcript_id \"ENST00000610020\"; saluki_score \"1.0\";  [92298965-92299146],[92300194-92300239],[92301476-92301590],[92303977-92304075],[92304284-92304349],[92307188-92307276],[92320599-92320634],[92323445-92324375],[92333391-92333473],[92336347-92336427],[92345846-92345914],[92380724-92380873],[92387011-92402056]\n",
      "3 + gene_id \"ENSG00000168314\"; transcript_id \"ENST00000311042\"; saluki_score \"1.0\";  [39467680-39467740],[39480040-39480123],[39502066-39502275],[39502463-39502876],[39513383-39516001]\n",
      "3 + gene_id \"ENSG00000185008\"; transcript_id \"ENST00000332191\"; saluki_score \"1.0\";  [77040730-77040846],[77098014-77098340],[77477414-77477571],[77481099-77481219],[77493244-77493382],[77522775-77522902],[77546338-77546462],[77550818-77550989],[77557944-77558149],[77562651-77562732],[77563167-77563329],[77564954-77565120],[77568313-77568434],[77574499-77574730],[77577490-77577614],[77579947-77580118],[77588751-77588933],[77595142-77595184],[77596623-77596750],[77602210-77602491],[77607798-77607954],[77617513-77617773],[77622227-77622432],[77632474-77632656],[77634870-77635043],[77644704-77644904],[77646054-77647116]\n",
      "3 - gene_id \"ENSG00000174963\"; transcript_id \"ENST00000525172\"; saluki_score \"1.0\";  [147386048-147388859],[147390931-147391246],[147395852-147396469],[147402728-147402812],[147403964-147404284]\n",
      "4 + gene_id \"ENSG00000126545\"; transcript_id \"ENST00000246891\"; saluki_score \"1.0\";  [69931081-69931117],[69932544-69932606],[69934212-69934244],[69934690-69934710],[69935926-69935949],[69936456-69936479],[69936566-69936607],[69937121-69937144],[69937800-69937823],[69939176-69939208],[69940021-69940044],[69941019-69941060],[69942046-69942063],[69942536-69942577],[69944850-69945004],[69946196-69946574]\n",
      "4 - gene_id \"ENSG00000164096\"; transcript_id \"ENST00000399075\"; saluki_score \"1.0\";  [119296419-119298779],[119300336-119300733],[119304245-119304445]\n",
      "4 + gene_id \"ENSG00000186352\"; transcript_id \"ENST00000335174\"; saluki_score \"1.0\";  [185396484-185396950],[185397150-185397302],[185398937-185399028],[185399570-185399773],[185400017-185400241]\n",
      "7 - gene_id \"ENSG00000048405\"; transcript_id \"ENST00000393313\"; saluki_score \"1.0\";  [127370043-127371814],[127373342-127375034],[127377186-127377329],[127386060-127386155],[127391497-127391615],[127392158-127392691]\n",
      "X + gene_id \"ENSG00000122824\"; transcript_id \"ENST00000376006\"; saluki_score \"1.0\";  [51332231-51332367],[51332883-51333459],[51336239-51337525]\n",
      "X - gene_id \"ENSG00000196368\"; transcript_id \"ENST00000375992\"; saluki_score \"1.0\";  [51490011-51491749],[51495951-51496596]\n",
      "8 + gene_id \"ENSG00000175445\"; transcript_id \"ENST00000311322\"; saluki_score \"1.0\";  [19938971-19939528],[19948180-19948340],[19951769-19951948],[19953310-19953421],[19954120-19954353],[19955841-19956083],[19959260-19959380],[19960901-19961083],[19962115-19962219],[19965310-19967258]\n",
      "8 + gene_id \"ENSG00000069206\"; transcript_id \"ENST00000380789\"; saluki_score \"1.0\";  [24441084-24441160],[24442473-24442576],[24447186-24447262],[24463882-24463960],[24465699-24465775],[24466799-24466988],[24468767-24468820],[24476433-24476504],[24482142-24482311],[24485277-24485361],[24487187-24487317],[24489159-24489333],[24490799-24490888],[24491903-24492098],[24492495-24492597],[24493043-24493229],[24499236-24499316],[24500178-24500256],[24500790-24500895],[24501477-24501576],[24506091-24506156],[24507480-24507535],[24508546-24509548]\n",
      "8 + gene_id \"ENSG00000104435\"; transcript_id \"ENST00000518111\"; saluki_score \"1.0\";  [79611152-79611214],[79636802-79636897],[79641378-79641550],[79654871-79655062],[79663580-79663661],[79664815-79665007]\n",
      "9 - gene_id \"ENSG00000122756\"; transcript_id \"ENST00000378980\"; saluki_score \"1.0\";  [34551432-34552071],[34552161-34552329],[34552674-34552854],[34556255-34556418],[34557526-34557692],[34557867-34557984],[34564599-34564832],[34568897-34568981],[34581095-34581205],[34589555-34589737]\n",
      "10 - gene_id \"ENSG00000203791\"; transcript_id \"ENST00000368836\"; saluki_score \"1.0\";  [124759954-124760503],[124762300-124762558],[124765392-124765608],[124774675-124774782],[124789043-124789157],[124790273-124790338],[124791724-124791870]\n",
      "13 - gene_id \"ENSG00000165480\"; transcript_id \"ENST00000314759\"; saluki_score \"1.0\";  [21153595-21155150],[21155693-21155811],[21157922-21158125],[21159902-21159987],[21161790-21161875],[21167988-21168399],[21172339-21172504],[21172620-21172681],[21176375-21176602]\n",
      "17 + gene_id \"ENSG00000108424\"; transcript_id \"ENST00000290158\"; saluki_score \"1.0\";  [47649838-47650284],[47650386-47650444],[47652694-47652876],[47656860-47657060],[47658508-47658660],[47661119-47661178],[47663089-47663178],[47664159-47664269],[47665057-47665158],[47668186-47668410],[47669678-47669869],[47670702-47670832],[47673018-47673165],[47673490-47673561],[47674638-47674782],[47676409-47676491],[47677020-47677127],[47678046-47678189],[47678308-47678413],[47680020-47680134],[47680508-47680669],[47682404-47685505]\n",
      "20 + gene_id \"ENSG00000254806\"; transcript_id \"ENST00000458187\"; saluki_score \"1.0\";  [45363502-45363693],[45365619-45365686],[45399690-45399751],[45408460-45408606],[45408801-45408938],[45409932-45409997]\n",
      "19 - gene_id \"ENSG00000196605\"; transcript_id \"ENST00000588267\"; saluki_score \"1.0\";  [9752140-9752440],[9758059-9758764],[9762082-9762165],[9763282-9763461],[9764810-9765035],[9768289-9768734]\n",
      "19 + gene_id \"ENSG00000188223\"; transcript_id \"ENST00000591613\"; saluki_score \"1.0\";  [35745678-35745700],[35745835-35745991],[35746419-35746523],[35752176-35752251],[35752434-35752484],[35752804-35752833],[35752914-35752999],[35753087-35753253],[35754017-35754157],[35754246-35754319],[35754393-35754519]\n",
      "22 - gene_id \"ENSG00000138944\"; transcript_id \"ENST00000381176\"; saluki_score \"1.0\";  [44243667-44249685],[44285428-44285745],[44296672-44296885],[44300879-44300977],[44312751-44312851]\n"
     ]
    }
   ],
   "source": [
    "########################################################\n",
    "# read gene data\n",
    "\n",
    "# open genes GTF\n",
    "genes_gtf = pygene.GTF(genes_gtf_file)\n",
    "\n",
    "# open genome FASTA\n",
    "genome_open = pysam.Fastafile(genome_fasta_file)\n",
    "\n",
    "# initialize dicts\n",
    "gene_seq1 = {}\n",
    "# gene_names = {}\n",
    "\n",
    "for gene_id, gene in genes_gtf.genes.items():\n",
    "    # choose top transcript\n",
    "    score_transcripts = [(tx.kv['saluki_score'],tx_id,tx) for tx_id, tx in gene.transcripts.items()]\n",
    "    _, tx_id, tx = sorted(score_transcripts)[-1]\n",
    "    \n",
    "    # one hot encode\n",
    "    tx_seq1 = rna_1hot_splice(tx, genome_open)\n",
    "    \n",
    "    # check coding\n",
    "    valid_coding = (tx_seq1[:,-2].sum() > 0)\n",
    "    \n",
    "    # save\n",
    "    if valid_coding:\n",
    "        gene_id = trim_dot(gene_id)\n",
    "        gene_seq1[gene_id] = tx_seq1\n",
    "        # gene_names[gene_id] = tx.kv['gene_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(12968, 0.9989985363223173)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter genes\n",
    "\n",
    "gene_mask = np.array([gene_id in gene_seq1 for gene_id in stability_df.Gene])\n",
    "gene_mask.sum(), gene_mask.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Fold             Gene  Stability\n",
      "522       2  ENSG00000048405  -1.724673\n",
      "2855      7  ENSG00000108424   0.868830\n",
      "4208      6  ENSG00000122484  -0.680720\n",
      "4238      6  ENSG00000122756   1.396266\n",
      "4243      1  ENSG00000122824  -0.200585\n",
      "8796      2  ENSG00000165480  -0.280310\n",
      "10178     2  ENSG00000174963  -0.002990\n",
      "10234     6  ENSG00000175445   0.429118\n",
      "11161     2  ENSG00000185008  -0.219503\n",
      "11333    10  ENSG00000186352  -1.703123\n",
      "11675     8  ENSG00000196368   0.057440\n",
      "11725     6  ENSG00000196605   0.607139\n",
      "12177     3  ENSG00000203791   0.883445\n"
     ]
    }
   ],
   "source": [
    "print(stability_df[~gene_mask])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(12968, 3)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter genes\n",
    "\n",
    "stability_df = stability_df[gene_mask].copy()\n",
    "stability_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "findfont: Font family ['sans-serif'] not found. Falling back to DejaVu Sans.\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Helvetica\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<seaborn.axisgrid.FacetGrid at 0x7f1ad0874f10>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjgAAAEYCAYAAABRMYxdAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAY50lEQVR4nO3dfbBcd33f8fcHY2wCoYCRHNlyxs5EwdgMmFY4OJ5pDKaJnDDIifEDQ4ggbjWdmqcmIVgwLk0zGjxDxwOlJalqMEpjMCqGkQKVjTE4TFpqI4wL2BZBIwlZ8a2vrECBpmMq8e0fe5Ss5b2rq6t7dveefb9mNLv7O+fs/Z44XH10fk+pKiRJkrrkaeMuQJIkabEZcCRJUucYcCRJUucYcCRJUucYcCRJUuc8fdwFnIg1a9bUHXfcMe4yJEnS+GRQ45J+gvP444+PuwRJkjSBlnTAkSRJGsSAI0mSOseAI0mSOseAI0mSOseAI0mSOseAI0mSOseAI0mSOseAI0mSOseAI0mSOseAI0mSOmdJ70UlScNctvYKZmYPDjy2YvlpbN96+4grkjQqBhxJnTUze5Bz3/y+gcd23rJhxNVIGiW7qCRJUucYcCRJUue02kWV5LnAzcCLgQJ+B/g28EngbGAvcFVVfa85fwNwLXAYeFtV3dlmfZKWtmFjbAD27tvHuSOsR9LkaHsMzgeBO6rqdUmeAfwU8G7g7qq6Mcn1wPXAu5KcB1wDnA+cAXwhyS9U1eGWa5S0RA0bYwOw64arR1iNpEnSWhdVkucA/xj4CEBV/biqvg+sBTY3p20GLm/erwVuq6onqmoPsAu4sK36JElSd7U5BufngAPALUm+nuTmJM8CTq+qGYDmdXlz/pnAI33X72/aniTJ+iQ7kuw4cOBAi+VLkqSlqs0uqqcD/xB4a1Xdm+SD9Lqj5pIBbfWUhqpNwCaA1atXP+W4pG4ZNs7GMTaS5tJmwNkP7K+qe5vPn6IXcB5LsqKqZpKsAGb7zj+r7/qVwKMt1idpCRg2zsYxNpLm0loXVVX9L+CRJC9smi4FHgK2AeuatnXA1ub9NuCaJKckOQdYBdzXVn2SJKm72p5F9Vbg1mYG1W7gzfRC1ZYk1wL7gCsBqurBJFvohaBDwHXOoJIkSQvRasCpqgeA1QMOXTrH+RuBjW3WJEmSus+VjCVJUucYcCRJUue4m7iksXMquKTFZsCRNHZOBZe02OyikiRJneMTHEmLZlhX04rlp7F96+0jrkjStDLgSFo0w7qadt6yYcTVSJpmdlFJkqTOMeBIkqTOMeBIkqTOcQyOpJHYs3s3F1x0ycBjrnUjabEZcCSNxGFwrRtJI2MXlSRJ6hwDjiRJ6hwDjiRJ6hwDjiRJ6hwDjiRJ6hwDjiRJ6hwDjiRJ6hwDjiRJ6hwDjiRJ6hxXMpY0lYZtHbFi+Wls33r7aAuStKgMOJKm0rCtI7bfcLXhR1riDDiSdJRh4WfnLRtGW4ykBXEMjiRJ6hwDjiRJ6hwDjiRJ6pxWA06SvUm+meSBJDuatucnuSvJd5rX5/WdvyHJriTfTvKrbdYmSZK6axRPcF5ZVRdU1erm8/XA3VW1Cri7+UyS84BrgPOBNcCHk5w0gvokSVLHjKOLai2wuXm/Gbi8r/22qnqiqvYAu4ALR1+eJEla6toOOAV8PsnXkqxv2k6vqhmA5nV5034m8EjftfubtidJsj7JjiQ7Dhw40GLpkiRpqWp7HZyLq+rRJMuBu5LsHHJuBrTVUxqqNgGbAFavXv2U45IkSa0GnKp6tHmdTfIZel1OjyVZUVUzSVYAs83p+4Gz+i5fCTzaZn2Sjt9la69gZvbgwGN79+3j3BHXI0mDtBZwkjwLeFpV/bB5/yvAvwG2AeuAG5vXrc0l24CPJ7kJOANYBdzXVn3StBsWVIZtRzAze3DOVX533XD1otUnSSeizSc4pwOfSXLk53y8qu5I8lVgS5JrgX3AlQBV9WCSLcBDwCHguqo63GJ90lQbFlTcjkDSUtdawKmq3cBLB7QfBC6d45qNwMa2apIkSdPBzTYlPcWe3bvn3E3bcTaSlgIDjqSnGLabtuNsJC0F7kUlSZI6xyc4knQchnXfDZt9Jmm0DDiSdByGdd85+0yaHHZRSZKkzjHgSJKkzjHgSJKkzjHgSJKkzjHgSJKkzjHgSJKkzjHgSJKkzjHgSJKkzjHgSJKkzjHgSJKkzjHgSJKkzjHgSJKkzjHgSJKkzjHgSJKkzjHgSJKkzjHgSJKkznn6uAuQ1J7L1l7BzOzBgcf27tvHuSOuR5JGxYAjddjM7EHOffP7Bh7bdcPVI65GkkbHLipJktQ5BhxJktQ5dlFJ0iLZs3s3F1x0ycBjK5afxvatt4+2IGmKGXAkaZEchjnHPG2/4WrDjzRCrQecJCcBO4C/rqrXJHk+8EngbGAvcFVVfa85dwNwLb3fE2+rqjvbrk+SRmFY+Nl5y4bRFiNNgVGMwXk78HDf5+uBu6tqFXB385kk5wHXAOcDa4APN+FIkiTpuLQacJKsBH4duLmveS2wuXm/Gbi8r/22qnqiqvYAu4AL26xPkiR1U9tPcD4A/AHwk76206tqBqB5Xd60nwk80nfe/qbtSZKsT7IjyY4DBw60UrQkSVraWgs4SV4DzFbV1+Z7yYC2ekpD1aaqWl1Vq5ctW3ZCNUqSpG5qc5DxxcBrk/wacCrwnCR/BjyWZEVVzSRZAcw25+8Hzuq7fiXwaIv1SZKkjmrtCU5VbaiqlVV1Nr3Bw1+sqt8CtgHrmtPWAVub99uAa5KckuQcYBVwX1v1SZKk7hrHOjg3AluSXAvsA64EqKoHk2wBHgIOAddV1eEx1CdJkpa4kQScqroHuKd5fxC4dI7zNgIbR1GTJEnqLveikiRJnWPAkSRJneNeVNISd9naK5iZPTjw2N59+zh3xPVI0iQw4EhL3MzswTn3ONp1w9UjrkaSJoNdVJIkqXN8giNJY7Zn924uuOiSgcdWLD+N7VtvH21BUgcYcCRpzA7DnN2MO2/ZMNpipI6wi0qSJHWOAUeSJHXOvAJOkovn0yZJkjQJ5vsE50PzbJMkSRq7oYOMk1wE/BKwLMnv9h16DnBSm4VJkiQt1LFmUT0DeHZz3k/3tf8AeF1bRUmSJJ2IoQGnqv4C+IskH6uq746oJkmSpBMy33VwTkmyCTi7/5qqelUbRUmSJJ2I+Qac/wL8CXAzvTWpJEmSJtZ8A86hqvrjViuRJElaJPOdJv7nSf5FkhVJnn/kT6uVSZIkLdB8n+Csa17f2ddWwM8tbjmSJEknbl4Bp6rOabsQSZKkxTKvgJPktwe1V9WfLm45kiRJJ26+XVQv73t/KnApcD9gwJEkSRNnvl1Ub+3/nOQfAP+5lYokSZJO0HxnUR3tb4FVi1mIJEnSYpnvGJw/pzdrCnqbbL4I2NJWUZKe7LK1VzAze3Dgsb379nHuiOvR6OzZvZsLLrpk4LEVy09j+9bbR1uQtETMdwzOv+17fwj4blXtb6EeSQPMzB7k3De/b+CxXTdcPeJqNEqHYc7/9jtv2TDaYqQlZF5dVM2mmzvp7Sj+PODHbRYlSZJ0IubbRXUV8H7gHiDAh5K8s6o+NeSaU4EvA6c0P+dTVfXeZgXkT9LbuHMvcFVVfa+5ZgNwLb1/tLytqu5c2G1Jk2dYNxPY3SBJi2m+XVTvAV5eVbMASZYBXwDmDDjAE8CrqupHSU4G/jLJduA3gbur6sYk1wPXA+9Kch5wDXA+cAbwhSS/UFVu7qlOGNbNBHY3SNJimu8sqqcdCTeNg8e6tnp+1Hw8uflTwFpgc9O+Gbi8eb8WuK2qnqiqPcAu4MJ51idJkvR35htw7khyZ5I3JXkT8Dngvx7roiQnJXkAmAXuqqp7gdOragageV3enH4m8Ejf5fubtqO/c32SHUl2HDhwYJ7lS5KkaTI04CT5+SQXV9U7gf8IvAR4KfAVYNOxvryqDlfVBcBK4MIkLx724wZ9xYDv3FRVq6tq9bJly45VgiRJmkLHeoLzAeCHAFX16ar63ar6l/Se3nxgvj+kqr5Pb4DyGuCxJCsAmtcjXV/7gbP6LlsJPDrfnyFJknTEsQLO2VX1jaMbq2oHvVlQc0qyLMlzm/fPBF5Nb6r5NmBdc9o6YGvzfhtwTZJTkpxDb6Xk++Z3G5IkSX/vWLOoTh1y7JnHuHYFsDnJSfSC1Jaq+mySrwBbklwL7AOuBKiqB5NsAR6it5jgdc6gkiRJC3GsgPPVJP+sqv5Tf2MTTr427MLmyc/LBrQfpLcb+aBrNgIbj1GTJAm3cZCGOVbAeQfwmSRv4O8DzWrgGcBvtFiXNHWG/WXlflMaxG0cpLkNDThV9RjwS0leCRyZAfW5qvpi65VJU2bYX1buNyVJx2deKxlX1ZeAL7VciyRJ0qKY70J/kiRJS4YBR5IkdY4BR5Ikdc58dxOXJC0hTiHXtDPgSFIHOYVc084uKkmS1DkGHEmS1DkGHEmS1DkGHEmS1DkGHEmS1DkGHEmS1DkGHEmS1DkGHEmS1DkGHEmS1DkGHEmS1DkGHEmS1DkGHEmS1DlutiktosvWXsHM7MGBx/bu28e5I65HkqaVAUdaRDOzB+fcwXnXDVePuBpJml52UUmSpM4x4EiSpM4x4EiSpM4x4EiSpM5xkLEkTZk9u3dzwUWXDDy2YvlpbN96+2gLklrQWsBJchbwp8DPAD8BNlXVB5M8H/gkcDawF7iqqr7XXLMBuBY4DLytqu5sqz5JmlaHYc7Zfjtv2TDaYqSWtNlFdQj4vap6EfAK4Lok5wHXA3dX1Srg7uYzzbFrgPOBNcCHk5zUYn2SJKmjWgs4VTVTVfc3738IPAycCawFNjenbQYub96vBW6rqieqag+wC7iwrfokSVJ3jWSQcZKzgZcB9wKnV9UM9EIQsLw57Uzgkb7L9jdtkiRJx6X1gJPk2cDtwDuq6gfDTh3QVgO+b32SHUl2HDhwYLHKlCRJHdJqwElyMr1wc2tVfbppfizJiub4CmC2ad8PnNV3+Urg0aO/s6o2VdXqqlq9bNmy9oqXJElLVpuzqAJ8BHi4qm7qO7QNWAfc2Lxu7Wv/eJKbgDOAVcB9bdUnLZQbaqrLhk0hB6eRa+locx2ci4E3At9M8kDT9m56wWZLkmuBfcCVAFX1YJItwEP0ZmBdV1WHW6xPWhA31FSXDZtCDk4j19LRWsCpqr9k8LgagEvnuGYjsLGtmiRJ0nRwqwZJktQ5BhxJktQ5BhxJktQ5BhxJktQ5BhxJktQ5bU4TlybasPVsXOtDkpY2A46m1rD1bFzrQxps2EKA/sNAk8SAIw0w7Je4qxVrmg1bCNB/GGiSGHCkAYb9Ene1YkmafA4yliRJnWPAkSRJnWPAkSRJnWPAkSRJnWPAkSRJnWPAkSRJnWPAkSRJneM6OOq0YdsxuGCfJHWXAUedNmw7Bhfsk6TusotKkiR1jgFHkiR1jl1UkqRF4U7jmiQGHEnSonCncU0SA44kqXU+3dGoGXAkSa3z6Y5GzUHGkiSpc3yCoyXPxfwkSUcz4GjJczE/SdLRWuuiSvLRJLNJvtXX9vwkdyX5TvP6vL5jG5LsSvLtJL/aVl2SJKn72hyD8zFgzVFt1wN3V9Uq4O7mM0nOA64Bzm+u+XCSk1qsTZIkdVhrAaeqvgz8zVHNa4HNzfvNwOV97bdV1RNVtQfYBVzYVm2SJKnbRj2L6vSqmgFoXpc37WcCj/Sdt79pkyRJOm6TMsg4A9pq4InJemA9wM/+7M+2WZMmiDOlJEnHY9QB57EkK6pqJskKYLZp3w+c1XfeSuDRQV9QVZuATQCrV68eGILUPc6UkiQdj1EHnG3AOuDG5nVrX/vHk9wEnAGsAu4bcW2SpDFwGwe1obWAk+QTwCXAC5LsB95LL9hsSXItsA+4EqCqHkyyBXgIOARcV1WH26pNkjQ53MZBbWgt4FTV6+c4dOkc528ENrZVjyRJmh6TMshYciCxJGnRGHA0MRxILOlojs/RQhlwJEkTy/E5WqhRL/QnSZLUOgOOJEnqHAOOJEnqHAOOJEnqHAOOJEnqHGdRaaRc60aSNAoGHI2Ua91IkkbBgKNF51MaSdK4GXC06HxKI2kUXOVYwxhwJElLkqscaxhnUUmSpM4x4EiSpM6xi0qSNFWGTYRw7E53GHAkSVNl2EQIx+50hwFHC+JUcEnSJDPgaEGcCi5JmmQGHElS5wxbI8enzNPBgKOBhnVBgb8gJE22YWvk+JR5OhhwNNCwLijwF4SkbnJ15O4w4EiS1HB15O5woT9JktQ5PsGZYk71liR1lQFnijnVW5LUVQacjvMpjSQtDgcgLy0GnA44VohZ895bBx7zKY0kzZ8DkJeWiQs4SdYAHwROAm6uqhvHXNJEMMRI0uTy6c7kmaiAk+Qk4D8A/wTYD3w1ybaqemi8lS2eYUHl0b9+hDPOPGvgMUOMJE2uYU93tt9w9ZzhBwxAbZmogANcCOyqqt0ASW4D1gJjCTgLDSMnElQc9CtJ3TIs/MDwALTQv2sWemyhYWvY35fjCnCpqpH/0LkkeR2wpqr+afP5jcAvVtVb+s5ZD6xvPr4Q+PbICx2fFwCPj7uIMZnmewfvf5rvf5rvHbz/ab7/+d7741W15ujGSXuCkwFtT0pgVbUJ2DSaciZLkh1VtXrcdYzDNN87eP/TfP/TfO/g/U/z/Z/ovU/aSsb7gf7nZiuBR8dUiyRJWqImLeB8FViV5JwkzwCuAbaNuSZJkrTETFQXVVUdSvIW4E5608Q/WlUPjrmsSTKVXXONab538P6n+f6n+d7B+5/m+z+he5+oQcaSJEmLYdK6qCRJkk6YAUeSJHWOAWcJSfJHSb6R5IEkn09yxrhrGqUk70+ys/m/wWeSPHfcNY1SkiuTPJjkJ0mmYtpokjVJvp1kV5Lrx13PKCX5aJLZJN8ady3jkOSsJF9K8nDz//dvH3dNo5Lk1CT3Jfmfzb3/4bhrGockJyX5epLPLuR6A87S8v6qeklVXQB8FvhXY65n1O4CXlxVLwH+Cpi23e2+Bfwm8OVxFzIKfVu3XAacB7w+yXnjrWqkPgY8ZfGyKXII+L2qehHwCuC6Kfrv/wTwqqp6KXABsCbJK8Zb0li8HXh4oRcbcJaQqvpB38dncdQiiF1XVZ+vqkPNx/9Bb52kqVFVD1fVNK3c/Xdbt1TVj4EjW7dMhar6MvA3465jXKpqpqrub97/kN5fdGeOt6rRqJ4fNR9Pbv5M1e/7JCuBXwduXuh3GHCWmCQbkzwCvIHpe4LT73eA7eMuQq06E3ik7/N+puQvOD1ZkrOBlwH3jrmUkWm6Zx4AZoG7qmpq7r3xAeAPgJ8s9AsMOBMmyReSfGvAn7UAVfWeqjoLuBV4y/BvW3qOdf/NOe+h9/h68K6lS9h87n+KHHPrFnVfkmcDtwPvOOopdqdV1eFmOMJK4MIkLx5zSSOT5DXAbFV97US+Z6IW+hNU1avneerHgc8B722xnJE71v0nWQe8Bri0OriI03H8958Gbt0y5ZKcTC/c3FpVnx53PeNQVd9Pcg+98VjTMuD8YuC1SX4NOBV4TpI/q6rfOp4v8QnOEpJkVd/H1wI7x1XLOCRZA7wLeG1V/e2461Hr3LpliiUJ8BHg4aq6adz1jFKSZUdmiSZ5JvBqpuj3fVVtqKqVVXU2vf/df/F4ww0YcJaaG5vuim8Av0JvhPk0+ffATwN3NVPl/2TcBY1Skt9Ish+4CPhckjvHXVObmgHlR7ZueRjYMk1btyT5BPAV4IVJ9ie5dtw1jdjFwBuBVzX/e3+g+Rf9NFgBfKn5Xf9VemNwFjRVepq5VYMkSeocn+BIkqTOMeBIkqTOMeBIkqTOMeBIkqTOMeBIkqTOMeBIGrkk72l2Sf5GM/33F5O8I8lPzePavUleMKD9nyf57eb9x5K8rnl/85FNGpO8e7HvRdJkcpq4pJFKchFwE3BJVT3RhJVnAP8dWF1Vjx/j+r3HOi/Jx4DPVtWnjmr/UVU9+wRvQdIS4BMcSaO2Ani8qp4AaILK64Az6C1u9iWAJH+cZEfzpOcPj/qOdya5r/nz8835/zrJ7x/9w5Lck2R1khuBZzZPjG5N8kdJ3t533sYkb2vnliWNmgFH0qh9HjgryV8l+XCSX66qf0dvn6lXVtUrm/PeU1WrgZcAv5zkJX3f8YOqupDe6tYfmM8Prarrgf9bVRdU1RvobQOwDiDJ0+gtCd+5DVylaWXAkTRSVfUj4B8B64EDwCeTvGnAqVcluR/4OnA+cF7fsU/0vV60wDr2AgeTvIze1idfr6qDC/kuSZPH3cQljVxVHQbuAe5J8k2aJylHJDkH+H3g5VX1vWZMzan9XzHH++N1M/Am4GeAj57A90iaMD7BkTRSSV6YZFVf0wXAd4Ef0ttMFeA5wP8B/neS04HLjvqaq/tev3IcP/7/JTm57/NngDXAy+lt6impI3yCI2nUng18KMlzgUPALnrdVa8HtieZqapXJvk68CCwG/hvR33HKUnupfePtNcfx8/eBHwjyf1V9Yaq+nEzqPn7zVMlSR3hNHFJU6sZXHw/cGVVfWfc9UhaPHZRSZpKzeJ/u4C7DTdS9/gER5IkdY5PcCRJUucYcCRJUucYcCRJUucYcCRJUucYcCRJUuf8f3pM7WnWalTaAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 576x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.displot(stability_df.Stability, aspect=2, height=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add length\n",
    "\n",
    "gene_lengths = np.array([gene_seq1[gene_id].shape[0] for gene_id in stability_df.Gene])\n",
    "stability_df['Length'] = gene_lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<seaborn.axisgrid.FacetGrid at 0x7f1ac7617550>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWAAAAFgCAYAAACFYaNMAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAaNUlEQVR4nO3de7Ccd33f8ffXAomLTSThY0k+smtxrKbYOFykuEUwFOKmaCiDgMRUHZpqWgd1BoOhlIBdZnqZjDruJCEEF5EoQKykgCsgVIIGg6NwMTMY4yUyx/Klli+Rjy0k2dQ1F8+xZX37xz6SV3tWOitpn/3t2X2/Zs7s7m+f55yvLudznvN7fpfITCRJ/XdG6QIkaVQZwJJUiAEsSYUYwJJUiAEsSYU8p3QBp2Pt2rV54403li5DkmYTnRrn9BXwo48+WroESTplczqAJWkuM4AlqRADWJIKMYAlqRADWJIKMYAlqZBaAzgiFkbEFyPi7oi4KyJeHRGLI+KmiLi3elzUcvw1EbEnIu6JiDfWWZsklVb3FfAfATdm5j8AXg7cBVwN7MzMlcDO6jURcRGwHrgYWAtsjoh5NdcnScXUFsAR8SLgdcCnATLzqcx8HFgHbK0O2wq8tXq+DrghM6cz8wFgD3BpXfVJUml1XgG/BDgI/FlE/G1EfCoiXggsycx9ANXjOdXx48BDLedPVW3HiIiNEXFbRNx28ODBGsuXpHrVGcDPAV4FfDIzXwn8nKq74Tg6zZWesV1HZm7JzNWZuXpsbKw3lUpSAXUG8BQwlZnfr15/kWYg74+IZQDV44GW489rOX858EiN9UlSUbWthpaZP46IhyLilzPzHuAy4M7qYwNwbfW4vTplB/C5iPgocC6wEri1rvqkUzE9PU2j0ZjRvmrVKhYsWFCgIs1ldS9H+V7gsxExH7gf+Nc0r7q3RcQVwF7gcoDM3B0R22gG9CHgysx8pub6pJPSaDS4avN2Fo5PHG17/OH7+Pi7Yc2aNQUr01xUawBn5i5gdYe3LjvO8ZuATXXWJJ2uheMTnD1xSekyNAScCSdJhRjAklSIASxJhczpPeGk09E+ouGpp54CYP78+ccc5wgH1cUA1shqH9Ewtes7POfMxSy98GVHj3GEg+pkAGsodBqf282Va+uIhscfvo/n/tJSRziobwxgDYX2q1mvXDUXGMAaGo7P1VzjKAhJKsQAlqRCDGBJKsQ+YA2lw4eeZnJycka7Y3o1SAxgDaUn9u/lugefZOn9z67z78gIDRoDWEPrrKUrHBWhgWYfsCQVYgBLUiEGsCQVYgBLUiEGsCQVYgBLUiEOQ5NOoH1Cx+TkJIcPFyxIQ8UAlk6gfULH1K6bWXThqsJVaVgYwJqT2hdgr/PKtHVCx+MP31fPF9FIMoA1J83cTqjclWmndSdcc0LdMIA1Z7VvJ1RKezeFa06oWwaw1AOuO6FT4TA0SSrEAJakQuyC0MA51S3mpbnGANbAcYt5jQoDWAPJLeY1CgxgDbxO42ydEqxhYABr4HXa380pwRoGBrDmhPZxtk4J1jBwGJokFWIAS1IhBrAkFWIAS1IhBrAkFWIAS1IhtQZwRDwYEZMRsSsibqvaFkfETRFxb/W4qOX4ayJiT0TcExFvrLM2SSqtH1fAb8jMV2Tm6ur11cDOzFwJ7KxeExEXAeuBi4G1wOaImNeH+iSpiBITMdYBr6+ebwW+BXy4ar8hM6eBByJiD3Ap8L0CNWoIucOxBk3dAZzANyIigT/JzC3AkszcB5CZ+yLinOrYceCWlnOnqjapJ9zhWIOm7gB+TWY+UoXsTRFx9wmOjQ5tOeOgiI3ARoDzzz+/N1VqZLjDsQZJrX3AmflI9XgA+DLNLoX9EbEMoHo8UB0+BZzXcvpy4JEOn3NLZq7OzNVjY2N1li9JtaotgCPihRFx1pHnwD8F7gB2ABuqwzYA26vnO4D1EbEgIlYAK4Fb66pPkkqrswtiCfDliDjydT6XmTdGxA+AbRFxBbAXuBwgM3dHxDbgTuAQcGVmPlNjfZJUVG0BnJn3Ay/v0P4YcNlxztkEbKqrJkkaJM6Ek6RCDGBJKsQAlqRC3JJIRU1PT9NoNI5pc4aaRoUBrKIajQZXbd7OwvGJo23OUNOoMIBV3MLxCTfc1EiyD1iSCjGAJakQA1iSCjGAJakQA1iSCjGAJakQA1iSCjGAJakQA1iSCnEmnPqqfe0H133QKDOA1Vftaz+47oNGmQGsvmtd+8F1HzTK7AOWpEIMYEkqxACWpEIMYEkqxACWpEIcBSH12OFDTzM5OXlM26pVq1iwYEGhijSoDGCpx57Yv5frHnySpfcH0Bxq9/F3w5o1awpXpkFjAEs1OGvpimP2uZM6sQ9YkgoxgCWpELsgpJp1uikH3piTASzVrv2mHHhjTk0GsNQH3pRTJ/YBS1IhBrAkFWIAS1IhBrAkFWIAS1IhBrAkFWIAS1IhBrAkFeJEDNVqenqaRqNx9PXk5CSHDxcsSBogBrBq1Wg0uGrzdhaOTwAwtetmFl24qnBV0mCovQsiIuZFxN9GxFer14sj4qaIuLd6XNRy7DURsSci7omIN9Zdm/pj4fgEZ09cwtkTl3Dm2HjpcqSB0Y8+4PcBd7W8vhrYmZkrgZ3VayLiImA9cDGwFtgcEfP6UJ8kFVFrAEfEcuCfAZ9qaV4HbK2ebwXe2tJ+Q2ZOZ+YDwB7g0jrrk6SS6r4C/hjwIaD1tsuSzNwHUD2eU7WPAw+1HDdVtR0jIjZGxG0RcdvBgwdrKVqS+qG2AI6INwMHMrMx68HVKR3ackZD5pbMXJ2Zq8fGxk6rRkkqqc5REK8B3hIRbwKeB7woIv4HsD8ilmXmvohYBhyojp8Czms5fznwSI31SVJRtV0BZ+Y1mbk8My+geXPtbzLzXwI7gA3VYRuA7dXzHcD6iFgQESuAlcCtddUnSaWVGAd8LbAtIq4A9gKXA2Tm7ojYBtwJHAKuzMxnCtQn1a7TPnHuETd6+hLAmfkt4FvV88eAy45z3CZgUz9qkkpq3yfOPeJGkzPhpELcJ04uxiNJhRjAklSIASxJhRjAklSIASxJhRjAklSIASxJhRjAklSIASxJhRjAklSIU5GlAeDiPKPJAJYGgIvzjCYDWBoQLs4zeuwDlqRCDGBJKsQAlqRCDGBJKsSbcOqZ6elpGo3GMW2Tk5McPlyoIGnAGcDqmUajwVWbt7NwfOJo29Sum1l04aqCVUmDywBWTy0cnzhmKNXjD99XsBppsNkHLEmFGMCSVIgBLEmFGMCSVIgBLEmFdBXAEfGabtokSd3r9gr4ui7bJEldOuE44Ih4NbAGGIuID7S89SJgXp2FSdKwm20ixnzgzOq4s1ranwB+s66iJGkUnDCAM/PbwLcj4vrM/Ls+1SRJI6HbqcgLImILcEHrOZn5a3UUJY26TnvEgfvEDZtuA/gLwB8DnwKeqa8cSTBzjzhwn7hh1G0AH8rMT9ZaiaRjuEfc8Ot2GNpXIuLdEbEsIhYf+ai1Mkkact1eAW+oHn+npS2Bl/S2HEkaHV0FcGauqLsQSRo1XQVwRPyrTu2Z+ee9LUeSRke3XRC/2vL8ecBlwA8BA1iSTlG3XRDvbX0dEb8E/EUtFUnSiDjV5Sh/Aaw80QER8byIuDUibo+I3RHxX6r2xRFxU0TcWz0uajnnmojYExH3RMQbT7E2SZoTuu0D/grNUQ/QXITnpcC2WU6bBn4tM38WEc8FvhsRXwPeDuzMzGsj4mrgauDDEXERsB64GDgX+OuI+PuZ6cQPSUOp2z7g3295fgj4u8ycOtEJmZnAz6qXz60+ElgHvL5q3wp8C/hw1X5DZk4DD0TEHuBS4Htd1qg+m56eptFoHH09OTnJ4cMFC5LmmG77gL8dEUt49mbcvd2cFxHzgAZwIfCJzPx+RCzJzH3V590XEedUh48Dt7ScPlW1tX/OjcBGgPPPP7+bMlSTRqPBVZu3s3B8AoCpXTez6MJVhauS5o5ud8R4B3ArcDnwDuD7ETHrcpSZ+UxmvgJYDlwaES870Zfp9Ck6fM4tmbk6M1ePjY11U75qtHB8grMnLuHsiUs4c2zGz0tJJ9BtF8RHgF/NzAMAETEG/DXwxW5OzszHI+JbwFpgf0Qsq65+lwEHqsOmgPNaTlsOPNJlfZI053Q7CuKMI+FbeWy2cyNiLCIWVs+fD/wT4G5gB89Obd4AbK+e7wDWR8SCiFhBc5TFrV3WJ0lzTrdXwDdGxNeBz1ev/znwV7OcswzYWvUDnwFsy8yvRsT3gG0RcQWwl2a3Bpm5OyK2AXfSvNF3pSMgJA2z2faEuxBYkpm/ExFvB15Ls6/2e8BnT3RuZv4IeGWH9sdozqTrdM4mYFN3pUvS3DZbF8THgJ8CZOZfZuYHMvPf0bz6/Vi9pUnScJstgC+ormSPkZm30dyeSJJ0imYL4Oed4L3n97IQSRo1swXwDyLiXe2N1Q20RofjJUldmm0UxPuBL0fEO3k2cFcD84G31ViXJA29EwZwZu4H1kTEG4Ajs9j+d2b+Te2VSdKQ63YtiG8C36y5FkkaKae6HrAk6TQZwJJUiAEsSYUYwJJUiAEsSYUYwJJUiAEsSYV0ux6wpMIOH3qaycnJY9pWrVrFggULClWk02UAS3PEE/v3ct2DT7L0/ub2iY8/fB8ffzesWbOmcGU6VQawuuY29OWdtXQFZ09cUroM9YgBrK65Db3UWwawTsqRbeih+SuwpFPnKAhJKsQAlqRCDGBJKsQAlqRCDGBJKsQAlqRCDGBJKsQAlqRCDGBJKsQAlqRCDGBJKsQAlqRCDGBJKsQAlqRCDGBJKsQAlqRCDGBJKsQAlqRCDGBJKsQ94aQ56vChp5mcnJzRvmrVKhYsWFCgIp0sA1iao57Yv5frHnySpffH0bbHH76Pj78b1qxZU7AydcsAluaws5auOLpLteae2vqAI+K8iPhmRNwVEbsj4n1V++KIuCki7q0eF7Wcc01E7ImIeyLijXXVJkmDoM6bcIeAf5+ZLwX+EXBlRFwEXA3szMyVwM7qNdV764GLgbXA5oiYV2N9klRUbQGcmfsy84fV858CdwHjwDpga3XYVuCt1fN1wA2ZOZ2ZDwB7gEvrqk+SSutLH3BEXAC8Evg+sCQz90EzpCPinOqwceCWltOmqrb2z7UR2Ahw/vnn11i1NPd0GhnhqIjBVXsAR8SZwJeA92fmExFx3EM7tOWMhswtwBaA1atXz3hfGmXtIyMcFTHYag3giHguzfD9bGb+ZdW8PyKWVVe/y4ADVfsUcF7L6cuBR+qsTxpGjoyYO2oL4Ghe6n4auCszP9ry1g5gA3Bt9bi9pf1zEfFR4FxgJXBrXfXpWNPT0zQajWPa/NVVqledV8CvAX4LmIyIXVXbf6AZvNsi4gpgL3A5QGbujohtwJ00R1BcmZnP1FifWjQaDa7avJ2F4xOAv7pK/VBbAGfmd+ncrwtw2XHO2QRsqqsmndjC8Ymjv7p2upkzOTnJ4cMlKpOGkzPh1FGnaa5Tu25m0YWrClYlDRcDWMfVfjPn8YfvK1iNNHxcjlKSCjGAJakQA1iSCjGAJakQA1iSCjGAJakQA1iSCjGAJakQA1iSCjGAJakQA1iSCjGAJakQA1iSCnE1tBHUafcL1/odTm7SOdgM4BHUvvsFuNbvsHKTzsFmAI+o1t0vwLV+h5mbdA4u+4AlqRADWJIKMYAlqRADWJIKMYAlqRADWJIKMYAlqRDHAY+A9plvznobXZ1mxoGz40oxgEdA+8w3Z72NrvaZceDsuJIM4BHROvPNWW+jzZlxg8M+YEkqxACWpEIMYEkqxACWpEIMYEkqxACWpEIMYEkqxACWpEKciCGNODfuLMcAHkKu/aCT4cad5RjAQ8i1H3SynJ5chgE8x7Vf7ULzivdFy17i2g/SgKstgCPiM8CbgQOZ+bKqbTHwP4ELgAeBd2Tm/63euwa4AngGuCozv15XbcOk/WoXvOKV5oo6R0FcD6xta7sa2JmZK4Gd1Wsi4iJgPXBxdc7miJhXY21D5chKZ0c+zhwbL12SpC7UFsCZ+R3gJ23N64Ct1fOtwFtb2m/IzOnMfADYA1xaV22SNAj6PQ54SWbuA6gez6nax4GHWo6bqtpmiIiNEXFbRNx28ODBWouVpDoNykSM6NCWnQ7MzC2ZuTozV4+NjdVcliTVp9+jIPZHxLLM3BcRy4ADVfsUcF7LccuBR/pcmyScmNFP/Q7gHcAG4NrqcXtL++ci4qPAucBK4NY+1yYJJ2b0U53D0D4PvB44OyKmgP9EM3i3RcQVwF7gcoDM3B0R24A7gUPAlZn5TF21SToxJ2b0R20BnJn/4jhvXXac4zcBm+qqR5IGzaDchJOkkWMAS1IhBrAkFeJiPAOu02I7DgmShoMBPODaF9txSJA0PAzgOeDIYjuShosBPMe0z1Jytwtp7jKA55j2WUqu/au6dZqaDN6L6AUDeA5qnaXkbheqW/sPffBeRK8YwJJm5dTkejgOWJIK8Qp4gBxvg01vsknDyQAeIG6wKY0WA3jAtI/59SabNLzsA5akQgxgSSrELghJJ81943rDAJZ00tw3rjcMYEmnxMkZp88AlnTaXC/i1BjAkk6b60WcGgO4oPaZb85601xml8TJM4ALap/55qw3abQYwIW1znxz1ps0WgzgPnGhHUntDOA+caEdSe0M4D5yoR1JrQxgSX3RqRtu1McJG8A1cYiZdKz2bjjHCRvAtXGImTRTezfcqDOAa+QQM42y9unJ/hY4kwEsqRbt05P9LXAmA1hSbVqnJ/tb4EwGcI94003SyTKAe8SbbpJOlgHcQ950k3QyDGBJA2PUJmsYwJKK6LSLxuTkJFu+cx+Lll8IDP9kDQO4C6P2U1nqh067aBy5dzIqkzUM4C6032D7yd57+Levn+SSS579T+KoB+nkte+i0X7vpNNV8jBd/BjAXWq/wXbdN3Z3/MktqXfar5KHrUti4AI4ItYCfwTMAz6VmdfW+fU6dS/A7D9lZ/vJLak3Wr/XTnX35UHtRhyoAI6IecAngF8HpoAfRMSOzLyzV1+j04SJ1k5/mNnFYPeCNBg69Ru3f78+9dRTAMyfP//oMe3f5526EdsDuR+hPVABDFwK7MnM+wEi4gZgHdCzAG40Gmz4yEd54YuXAfDo/XewcMWxHf6/+Ml+/uuf38uiZXccc8wZZzT/0X928GGe8+STPHrmC4+e09422+teHWMt1jJytZy5mFadvl/nPf8sFi37e0ePaf8+bz/n54/t44Prf33GfZ3fv+Gmo1nx88f2sXXTB3ra/RGZ2bNPdroi4jeBtZn529Xr3wL+YWa+p+WYjcDG6uUvA/f0vdDjOxt4tHQRs7DG3rDG3hiVGh/NzLXtjYN2BRwd2o75CZGZW4At/Snn5ETEbZm5unQdJ2KNvWGNvTHqNZ5Rxyc9DVPAeS2vlwOPFKpFkmo1aAH8A2BlRKyIiPnAemBH4ZokqRYD1QWRmYci4j3A12kOQ/tMZu4uXNbJGMiukTbW2BvW2BsjXeNA3YSTpFEyaF0QkjQyDGBJKsQA7pGIeF9E3BERuyPi/aXrAYiIz0TEgYi4o6VtcUTcFBH3Vo+LBrDGy6u/x8MRUXyI0nFq/L2IuDsifhQRX46IhQVLPF6Nv1vVtysivhER5w5ajS3vfTAiMiLOLlFbSx2d/h7/c0Q8XP097oqIN/Xq6xnAPRARLwPeRXMm38uBN0fEyrJVAXA90D74+2pgZ2auBHZWr0u6npk13gG8HfhO36vp7Hpm1ngT8LLM/BXg/wDX9LuoNtczs8bfy8xfycxXAF8F/mO/i2pzPTNrJCLOo7n8wN5+F9TB9XSoEfjDzHxF9fFXvfpiBnBvvBS4JTN/kZmHgG8DbytcE5n5HeAnbc3rgK3V863AW/tZU7tONWbmXZk5MDMcj1PjN6p/a4BbaI5ZL+Y4NT7R8vKFtE1q6rfj/H8E+EPgQxSuD05YYy0M4N64A3hdRLw4Il4AvIljJ5QMkiWZuQ+gejyncD3D4N8AXytdRCcRsSkiHgLeSfkr4Bki4i3Aw5l5e+laZvGeqjvnM73stjOAeyAz7wL+G81fS28EbgcOnfAkDYWI+AjNf+vPlq6lk8z8SGaeR7O+98x2fD9VFysfYQB/MLT5JDABvALYB/xBrz6xAdwjmfnpzHxVZr6O5q8w95au6Tj2R8QygOrxQOF65qyI2AC8GXhnDv6A+s8Bv1G6iDYTwArg9oh4kGY3zg8jYmnRqtpk5v7MfCYzDwN/SvNeT08YwD0SEedUj+fTvIH0+bIVHdcOYEP1fAOwvWAtc1a1ccCHgbdk5i9K19NJ243gtwB3l6qlk8yczMxzMvOCzLyA5lowr8rMHxcu7RhHLlgqb6PZ5dgbmelHDz6Am2muW3w7cFnpeqqaPk/zV6anaf7nvgJ4Mc3RD/dWj4sHsMa3Vc+ngf3A1wewxj3AQ8Cu6uOPB7DGL1Vh8SPgK8D4oNXY9v6DwNmDViPwF8Bk9fe4A1jWq6/nVGRJKsQuCEkqxACWpEIMYEkqxACWpEIMYEkqxADWUImIn9X8+d9fzeDqy9fTcDOApZPzfuAFsx0kdWOg9oST6hARE8AngDHgF8C7MvPuiLgeeAJYDSwFPpSZX4yIM4D/Dvxj4AGaFyqfAc6tPr4ZEY9m5huqz7+J5pTkJ4F1mbm/n38+zV1eAWsUbAHem5mrgA8Cm1veWwa8lmaAXlu1vR24ALgE+G3g1QCZ+XHgEeANR8KX5jKPt2Tmy2muX/yuWv8kGipeAWuoRcSZwBrgCxFxpHlByyH/K5uLrNwZEUuqttcCX6jafxwR3zzBl3iK5mLnAA2aC4tLXTGANezOAB7P5q4QnUy3PI+2x248nc/O538Gv6d0EuyC0FDL5q4QD0TE5QDR9PJZTvsu8BsRcUZ1Vfz6lvd+CpxVS7EaOQawhs0LImKq5eMDNHeDuCIibgd209yW6US+RHMlrDuAPwG+D/y/6r0twNdm6ZaQuuJqaFIHEXFmZv4sIl4M3Aq8JgdsnVrNffZXSZ19tdpqfj7wu4av6uAVsCQVYh+wJBViAEtSIQawJBViAEtSIQawJBXy/wF/L+y9gAumvgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 360x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.displot(np.log2(1+stability_df.Length))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(16384, 14.0, 0.9966070326958667)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_len_t = 2**14\n",
    "full_len_t, np.log2(full_len_t), (stability_df.Length<full_len_t).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(12288, 13.584962500721156, 0.9844231955582974)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_len_t = 3*2**12\n",
    "full_len_t, np.log2(full_len_t), (stability_df.Length<full_len_t).mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Training data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set fold masks\n",
    "\n",
    "num_folds = len(set(stability_df.Fold))\n",
    "fold_masks = []\n",
    "for fi in range(num_folds):\n",
    "    fold_indexes = np.where(stability_df.Fold == fi+1)[0]\n",
    "    fold_mask = np.zeros(stability_df.shape[0], dtype='bool')\n",
    "    fold_mask[fold_indexes] = 1\n",
    "    fold_masks.append(fold_mask)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def feature_bytes(values):\n",
    "    \"\"\"Convert numpy arrays to bytes features.\"\"\"\n",
    "    values = values.flatten().tostring()\n",
    "    return tf.train.Feature(bytes_list=tf.train.BytesList(value=[values]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tfr_feature_ints(int_list):\n",
    "    return tf.train.Feature(int64_list=tf.train.Int64List(value=int_list))\n",
    "def tfr_feature_dna(seq_1hot):\n",
    "    seq_indexes = np.where(seq_1hot)[1].astype('uint8')\n",
    "    seq_indexes = seq_indexes.tostring()\n",
    "    return tf.train.Feature(bytes_list=tf.train.BytesList(value=[seq_indexes]))\n",
    "def tfr_feature_track(seq_track):\n",
    "    seq_track = seq_track.astype('uint8')\n",
    "    seq_track = seq_track.tostring()\n",
    "    return tf.train.Feature(bytes_list=tf.train.BytesList(value=[seq_track]))\n",
    "def tfr_feature_targets(targets):\n",
    "    targets = targets.astype('float16')\n",
    "    targets = targets.tostring()\n",
    "    return tf.train.Feature(bytes_list=tf.train.BytesList(value=[targets]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_fold(fold_index, out_dir, seqs_per_tfr=128):\n",
    "    tf_opts = tf.io.TFRecordOptions(compression_type='ZLIB')\n",
    "    \n",
    "    # slice fold's genes\n",
    "    fold_df = stability_df[fold_masks[fold_index]]\n",
    "    \n",
    "    gi = 0\n",
    "    num_tfr = int(np.ceil(fold_df.shape[0] / seqs_per_tfr))\n",
    "    for fi in range(num_tfr):\n",
    "        tfr_file = '%s/fold%d-%d.tfr' % (out_dir,fold_index,fi)\n",
    "        with tf.io.TFRecordWriter(tfr_file, tf_opts) as writer:\n",
    "            for ti in range(seqs_per_tfr):\n",
    "                if gi < fold_df.shape[0]:\n",
    "                    sgene = fold_df.iloc[gi]\n",
    "                    gene_id = sgene.Gene\n",
    "                    \n",
    "                    # trim sequence\n",
    "                    seq1 = gene_seq1[gene_id][:full_len_t]\n",
    "                    seq_len = seq1.shape[0]\n",
    "                    \n",
    "                    # split\n",
    "                    sequence = seq1[:,:4]\n",
    "                    coding = seq1[:,4]\n",
    "                    splice = seq1[:,5]\n",
    "                        \n",
    "                    # targets\n",
    "                    # targets = np.log2(sgene.Stability + pseudocount)\n",
    "                    targets = sgene.Stability\n",
    "\n",
    "                    # hash to bytes\n",
    "                    features_dict = {\n",
    "                        'lengths': tfr_feature_ints([seq_len]),\n",
    "                        'sequence': tfr_feature_dna(sequence),\n",
    "                        'coding': tfr_feature_track(coding),\n",
    "                        'splice': tfr_feature_track(splice),\n",
    "                        'targets': tfr_feature_targets(targets)\n",
    "                    }\n",
    "\n",
    "                    # write example\n",
    "                    example = tf.train.Example(features=tf.train.Features(feature=features_dict))\n",
    "                    writer.write(example.SerializeToString())\n",
    "                \n",
    "                    # next gene\n",
    "                    gi += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_8716/1940088090.py:5: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  seq_indexes = seq_indexes.tostring()\n",
      "/tmp/ipykernel_8716/1940088090.py:9: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  seq_track = seq_track.tostring()\n",
      "/tmp/ipykernel_8716/1940088090.py:13: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.\n",
      "  targets = targets.tostring()\n"
     ]
    }
   ],
   "source": [
    "################################################################\n",
    "# output directory\n",
    "\n",
    "out_dir = 'data'\n",
    "if os.path.isdir(out_dir):\n",
    "    shutil.rmtree(out_dir)\n",
    "os.mkdir(out_dir)\n",
    "\n",
    "################################################################\n",
    "# stats\n",
    "stats_dict = {}\n",
    "stats_dict['num_targets'] = 1\n",
    "stats_dict['target_length'] = 1\n",
    "stats_dict['length_t'] = full_len_t\n",
    "\n",
    "stats_dict['num_folds'] = num_folds\n",
    "for fi in range(num_folds):\n",
    "    stats_dict['fold%d_seqs' % fi] = int(fold_masks[fi].sum())\n",
    "\n",
    "with open('%s/statistics.json' % out_dir, 'w') as stats_json_out:\n",
    "    json.dump(stats_dict, stats_json_out, indent=4)\n",
    "    \n",
    "################################################################\n",
    "# write gene table\n",
    "\n",
    "data_genes_df_list = [] \n",
    "for fi in range(num_folds):\n",
    "    fold_df = stability_df[fold_masks[fi]]    \n",
    "    data_genes_df_list.append(fold_df)\n",
    "\n",
    "data_genes_df = pd.concat(data_genes_df_list, axis=0)\n",
    "data_genes_df.set_index(np.arange(data_genes_df.shape[0]), inplace=True)\n",
    "data_genes_df['Fold'] = data_genes_df.Fold - 1\n",
    "data_genes_df.to_csv('%s/genes.tsv' % out_dir, sep='\\t')\n",
    "\n",
    "################################################################\n",
    "# write tfrecords\n",
    "\n",
    "tfrecords_dir = '%s/tfrecords' % out_dir\n",
    "os.mkdir(tfrecords_dir)\n",
    "for fi in range(num_folds):\n",
    "    write_fold(fi, tfrecords_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
